front/app/components/FormBuilder/components/utils.ts

```diff
@@ -0,0 +1,25 @@
+import { IFlatCustomField } from 'services/formCustomFields';
+
+type NumberHash = Record<string, number>;
+
+export const getFieldNumbers = (formCustomFields: IFlatCustomField[]) => {
```

Comment:
This is the main fix: we now calculate the field numbers, and put them in an object where the key is the field id and the value is the number grouped by `input_type`. This replaces `getIndexForTitle`. It's still not perfect but I personally found this easier to reason about. Plus now we're doing a lot less work on every re-render, since we only have to loop over the fields once instead of multiple times in various nested loops